pkg/buildermgr/fnwatcher.go

```diff
@@ -0,0 +1,158 @@
+/*
+Copyright 2017 The Fission Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package buildermgr
+
+import (
+	"time"
+
+	"go.uber.org/zap"
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/kubernetes"
+	k8sCache "k8s.io/client-go/tools/cache"
+
+	fv1 "github.com/fission/fission/pkg/apis/core/v1"
+	"github.com/fission/fission/pkg/crd"
+)
+
+type (
+	fnWatcher struct {
+		logger        *zap.Logger
+		fissionClient *crd.FissionClient
+		k8sClient     *kubernetes.Clientset
+	}
+)
+
+func makefnWatcher(logger *zap.Logger, fissionClient *crd.FissionClient,
+	k8sClientSet *kubernetes.Clientset) *fnWatcher {
+
+	fnw := &fnWatcher{
+		logger:        logger.Named("function_watcher"),
+		fissionClient: fissionClient,
+		k8sClient:     k8sClientSet,
+	}
+	return fnw
+}
+
+func (fnw *fnWatcher) watchFunctions() {
+	lw := k8sCache.NewListWatchFromClient(fnw.fissionClient.CoreV1().RESTClient(), "functions", apiv1.NamespaceAll, fields.Everything())
+
+	// TODO: change resyncperiod duration to the optimum value
+	_, controller := k8sCache.NewInformer(lw, &fv1.Function{}, 1*time.Minute, k8sCache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			fn := obj.(*fv1.Function)
+			fnw.updateFuncStatus(fn)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			fn := newObj.(*fv1.Function)
+			fnw.updateFuncStatus(fn)
+		},
+	})
+	controller.Run(make(chan struct{}))
+}
+
+func (fnw *fnWatcher) updateFuncStatus(fn *fv1.Function) {
+	fnw.pkgStatus(fn)
+
+	fnw.envStatus(fn)
+
+	if fv1.BuildStatusSucceeded == fn.Status.PackageStatus &&
+		fv1.BuildStatusSucceeded == fn.Status.EnvStatus {
+		fn.Status.FnStatus = fv1.BuildStatusSucceeded
+	} else if fv1.BuildStatusFailed == fn.Status.PackageStatus ||
+		fv1.BuildStatusFailed == fn.Status.EnvStatus {
+		fn.Status.FnStatus = fv1.BuildStatusFailed
+	} else {
+		fn.Status.FnStatus = fv1.BuildStatusPending
+	}
```

Comment:
I will think through this again refactor the code. Also, could you please point out some scenarios where - call to `fnw.updateFuncStatus(fn)` may break? It will help me improve the design.